### PR TITLE
Add an upload-blob subcommand and new sget client.

### PR DESCRIPTION
**Pull request by dlorenc**
Ref #178

### CODE REVIEW

**Comment by lukehinds** [comment]
This might be a good candidate for sigstore/sigstore?

**Comment by dlorenc** [comment]
> This might be a good candidate for sigstore/sigstore?

Do you want to put OCI registry interactions in there?

File: cmd/cosign/cli/upload_blob.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+				return flag.ErrHelp
+			}
+
+			return UploadBlobCmd(ctx, *fileRef, *ct, args[0])
+		},
+	}
+}
+
+func UploadBlobCmd(ctx context.Context, fileRef, contentType, imageRef string) error {
+
```
Comment (jonjohnsonjr) [inline cmd/cosign/cli/upload_blob.go:57]: extra newline

File: pkg/cosign/remote/remote.go [hunk 2]
```diff
@@ -163,38 +169,57 @@ func Upload(ctx context.Context, signature, payload []byte, dst name.Reference,
...
+	return img, nil
+}
+
+type StaticLayer struct {
 	b  []byte
 	mt types.MediaType
 }
 
-func (l *staticLayer) Digest() (v1.Hash, error) {
+func (l *StaticLayer) Digest() (v1.Hash, error) {
```
Comment (jonjohnsonjr) [inline pkg/cosign/remote/remote.go:196]: Maybe something worth upstreaming

File: cmd/cosign/cli/upload_blob.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+func UploadBlob() *ffcli.Command {
+	var (
+		flagset = flag.NewFlagSet("cosign upload-blob", flag.ExitOnError)
+		fileRef = flagset.String("f", "", "reference to the file to upload")
+		ct      = flagset.String("ct", "", "content type to set")
+	)
+	return &ffcli.Command{
+		Name:       "upload-blob",
+		ShortUsage: "cosign upload-blob -f <blob ref> <image uri>",
+		ShortHelp:  "upload a blob to the supplied container image",
```
Comment (jonjohnsonjr) [inline cmd/cosign/cli/upload_blob.go:44]: `to the supplied container image` is a bit weird, I'm not sure what it means

File: cmd/cosign/cli/upload_blob.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		return err
+	}
+
+	b, err := ioutil.ReadFile(fileRef)
+	if err != nil {
+		return err
+	}
+
+	if contentType == "" {
+		contentType = http.DetectContentType(b)
```
Comment (jonjohnsonjr) [inline cmd/cosign/cli/upload_blob.go:69]: neat

File: cmd/cosign/cli/upload_blob.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func UploadBlobCmd(ctx context.Context, fileRef, contentType, imageRef string) error {
+
+	ref, err := name.ParseReference(imageRef)
+	if err != nil {
+		return err
+	}
+
+	b, err := ioutil.ReadFile(fileRef)
```
Comment (jonjohnsonjr) [inline cmd/cosign/cli/upload_blob.go:63]: Probably don't want to buffer this.

File: cmd/cosign/cli/upload_blob.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		contentType = http.DetectContentType(b)
+	}
+
+	fmt.Fprintf(os.Stderr, "Uploading file from [%s] to [%s] with media type [%s]\n", fileRef, ref.Name(), contentType)
+	img, err := cremote.Upload(b, ref, types.MediaType(contentType), authn.DefaultKeychain)
+	if err != nil {
+		return err
+	}
+	sha := sha256.Sum256(b)
+	blobURL := ref.Context().Registry.RegistryStr() + "/v2/" + ref.Context().RepositoryStr() + "/blobs/sha256:" + hex.EncodeToString(sha[:])
```
Comment (jonjohnsonjr) [inline cmd/cosign/cli/upload_blob.go:78]: Wish there was a better way to do this but I don't have any ideas.

**Comment by dlorenc** [comment]
Now with multi-arch!

multi

File: cmd/cosign/cli/upload_blob.go (hunk 6 above)
Comment (dekkagaijin) [reply cmd/cosign/cli/upload_blob.go]: what's wrong with it, other than being a long read?

Comment (dekkagaijin) [reply cmd/cosign/cli/upload_blob.go]: I could whip up another set of `Reference`s for ya. It'll involve a lot of copy-paste

File: README.md [hunk 7]
```diff
@@ -121,6 +121,74 @@ The following checks were performed on these signatures:
...
+
+Verification for us.gcr.io/dlorenc-vmtest2/readme --
+The following checks were performed on each of these signatures:
+  - The cosign claims were validated
+  - Existence of the claims in the transparency log was verified offline
+  - The signatures were verified against the specified public key
+  - Any certificates were verified against the Fulcio roots.
+```
+
+The signature, claims and transparency log proofs are all verified automatically by sget as part of the download.
```
Comment (priyawadhwa) [inline README.md:188]: 🔥

File: cmd/cosign/cli/upload_blob.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	dgst, err := dgster.Digest()
+	if err != nil {
+		return err
+	}
+	dgstAddr := fmt.Sprintf("%s@%s", ref.Context().Name(), dgst.String())
+
+	if len(files) > 1 {
+		fmt.Fprintf(os.Stderr, "Uploading multi-platform index to %s\n", dgstAddr)
+	} else {
+		fmt.Fprintf(os.Stderr, "Uploaded image to to %s\n", dgstAddr)
```
Comment (priyawadhwa) [inline cmd/cosign/cli/upload_blob.go:128]: nit: to to

File: cmd/cosign/cli/upload_blob.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+func (fs *Files) Set(k string) error {
+	f := fileFromFlag(k)
+	fs.Files = append(fs.Files, f)
+
+	// If we have multiple files, each file must have a platform.
+	if len(fs.Files) > 1 {
+		for _, f := range fs.Files {
+			if f.Platform == nil {
+				return fmt.Errorf("each file must include a unique platform, %s had no platform", f.Path)
```
Comment (priyawadhwa) [inline cmd/cosign/cli/upload_blob.go:57]: do we check somewhere that each platform is also unique?

**Review by priyawadhwa — APPROVED** [review]
LGTM! 

You could add an integration test for `cosign upload-blob` with `sget`  in [e2e_test_secrets.sh](https://github.com/sigstore/cosign/blob/main/test/e2e_test_secrets.sh) just to make sure `sget` is working, but doesn't have to be in this PR!

File: README.md [hunk 10]
```diff
@@ -121,6 +121,74 @@ The following checks were performed on these signatures:
...
+
+`curl | bash` isn't a great idea, but `sget | bash` is slightly less-bad.
```
Comment (dekkagaijin) [inline README.md:190]: `slightly` -> `distinctly`
Gotta own that, this would (could?) have prevented CodeCov

File: test/e2e_test.go [hunk 11]
```diff
@@ -421,6 +423,56 @@ func TestUploadDownload(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	dgst, err := img.Digest()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// But pass by digest
+	rc, err := sget.SgetCmd(ctx, imgName+"@"+dgst.String(), "")
```
Comment (priyawadhwa) [inline test/e2e_test.go:463]: Yay!

**Review by priyawadhwa — APPROVED** [review]
Woot!

**Pull request merged**